examples/rollup.config.js

```diff
@@ -1,14 +1,24 @@
-import resolve from "@rollup/plugin-node-resolve";
+import date from 'date-and-time';
+import * as fs from 'node:fs';
+import fse from 'fs-extra';
+import { fileURLToPath } from 'node:url';
+import path from 'node:path';
+
+// 1st party Rollup plugins
+import alias from '@rollup/plugin-alias';
 import commonjs from "@rollup/plugin-commonjs";
 import replace from '@rollup/plugin-replace';
-import typescript from 'rollup-plugin-typescript2';
+import resolve from "@rollup/plugin-node-resolve";
 import terser from '@rollup/plugin-terser';
-import alias from '@rollup/plugin-alias';
+
+// 3rd party Rollup plugins
 import sourcemaps from 'rollup-plugin-sourcemaps';
-import date from 'date-and-time';
-import path from 'path';
-import fs from 'fs';
-import fse from 'fs-extra';
+import typescript from 'rollup-plugin-typescript2';
```

Comment:
We should switch to the official typescript plug-in at some point

Reply:
Yes. I'm preparing a PR but it still breaks sourcemaps or produces very large ones.  The issue is the configuration is not consistent.

Quick question that can resolve it.  Do the examples javascript bundles need to all be umd ? Can they all be ESM ?

This builds 3 javascript bundles, 2 in ESM, one in UMD.  In this configuration typescript is doing the work to make them es5 instead of rollup.  If I had a direction on the outcome, I could make this build system consistent.

Reply:
These are the primary goals for the examples, but I'm not entirely sure how to get there (not my area of expertise):
- the result needs to be ES5 so that we can run them on old devices / browsers
- in debug build, we need source maps so that we can see the engine source code in the original form

At the moment the examples are built using ES5 engine build (unless you specify the module build on command line). The reason is that we could not get all examples to work due to some internal reasons (for now) using ES6 engine. See these two issues for the reasons we're aware of:
https://github.com/playcanvas/engine/issues/5399
https://github.com/playcanvas/engine/issues/5384

Thanks for doing an awesome job on this! Much appreciated.

Reply:
That's what I need, thank you.  

Are we married to this build system ?  I could replace this with `vite` and 8 lines of config for a great dev experience.

In a way where you'll still have all the same utility (library swapping) and equivalent build products. It uses all the same tools under the hood, provides dev web server, hmr, consistency. I understand going rollup for the smallest, most treeshakable code for the engine library builds.  For everything else, cruise control.

If for any other reason, just being able to build without stopping the web server or having to mystery processes running after termination.

Reply:
Vite sounds interesting, and we might consider it, but we have many repos (internal / external) that use rollup, and for simplicity it's useful for us to keep the builds consistent - otherwise we find it hard to maintain.

But I'd be curious to see what @willeastcott has to say on this.

Reply:
Cruise control would be 100% build step free - nothing to worry about at all during the dev cycle :wink: And a plain ES5/ES6 output once the dust has settled.

Reply:
Vite is like "managed rollup", definitely keep fine grained rollup for specific build products.  For most of the development time, its build step free.

Reply:
To be clear, I'm just talking about "examples" project. Only to polish the front end heavy or "new to library" developer experience.  Get them in and going with training wheels and a soft chair.

Reply:
Please excuse all the messages: 

one last thing, the examples are built using React 18.  That does not support IE 11. (see: https://react.dev/blog/2022/03/08/react-18-upgrade-guide#dropping-support-for-internet-explorer)

In short, since march 31, examples browser works on the lowest supported target of React 18.  If you haven't had complaints, nobody is requiring es5 for examples, certainly not for old browser support.

Reply:
Some thoughts/comments from me:

The example browser has never supported IE11 since PCUI doesn't support IE11 (I believe because of it's reliance on flexbox and maybe some other CSS features). There's a mode that you can run examples from the examples browser outside of the application and that's how we used to test in IE11, IIRC. But we've pretty much given up supporting IE11, even in the engine codebase, now.

On Vite, as @mvaligursky says, sounds interesting. I just had a quick look at their site. But my immediate concern is just introducing another technology which folks have to learn. Right now, once you know Rollup, can understand straight away how our repos are built. I'm not averse to switching to other build systems, but my preference would be that we keep things consistent. Also, I want to make sure we examine all the options. One thing I've been thinking about recently (and it seems @kungfooman has too) is whether ESM examples + import maps would be a better way to go. Dump the TypeScript mode. Remove the build system for the examples browser. Remove the React stuff and use vanilla PCUI. And just make it more simple. This is how the PCUI examples work and I _much_ prefer developing/debugging against those:

https://github.com/playcanvas/pcui/tree/main/examples

Reply:
Yes, you are very right. This can be set up so examples can be run in browser or bundler without modification.  @mvaligursky invited me to comment, I got excited and wrote too much.

See: https://github.com/playcanvas/engine/issues/5541

Adjusting the examples that way is a great mark of quality. So is keeping typescript and doing the same for it. ESM + TS presented in a way that works the same bundle to browser is like having a mini playground / validation tool. It would be impressive.

If you feel adventurous, invite me to demonstrate it on something like `spine` that doesn't have a build system.  You'll see its as familiar as rollup (cause it is, check rollups front page) with a managed config.

Reply:
I'm pretty much averse to anything that I feel is unnecessary extra steps (like `npm build`, `npm serve`, TS sugar, JSX sugar, ...)

If every path is so hardcoded that you need `npm serve` or otherwise a user can't even open a simple `index.html` via lighttpd on `/var/www` then I think something is wrong and not resilient enough, so I ended up changing multiple things I required at once.

I could make a PR with what I have, but it requires more work (there are many examples left that need to be reworked from TSX to ESM), even though I invested quite some time already to make the process work in the first place.

If we could put this on a branch and more people are willing to contribute to that, we could work it out over time, without asking too much from one individual.

I'm not advocating for completely removing the examples build system (it's fine to create minimized bundles and we need thumbnails generation), but to only use what modern browser technology is giving us to enable a fast and fun dev experience.

Reply:
All those goals can be met. PR's incoming.

Reply:
@epreston Nice, what's your plan? Can you break it down and make an issue for the first PR? First step should be imo `*.tsx -> *.jsx` and `*.ts -> *.mjs` (too keep the amount of changes reviewable and everything intact)

Reply:
- Fix all the things external to engine blocking build improvements.
- Remove the workarounds in place to consume legacy code
- Predictable and easy to rationalise development
- Learn and share